Implement PHP support

Jeckerson opened the pull request.

Jeckerson committed:
Implement support for PHP

Jeckerson committed:
Remove trailing backslash

Jeckerson committed:
Improve PHP support

Jeckerson committed:
Adapt generation of PHP code

Jeckerson committed:
Adjust strategy with `$` prefix in variables

Jeckerson committed:
Update code of `01_basic.php` example

Jeckerson committed:
Improve PHP support

Jeckerson committed:
Change 're2c:yych:literals' to 'char-or-hex'

Jeckerson committed:
Adjust parse_u32_blocks.php example

Jeckerson committed:
Regenerate 01_basic.php

Jeckerson committed:
Regenerate parse_u32_conditions.php

Jeckerson committed:
Regenerate unicode_identifier.php

Jeckerson committed:
Regenerate examples/php/conditions/ examples

Jeckerson committed:
Regenerate 01_sentinel.php

Jeckerson committed:
Regenerate eof/ directory

Jeckerson committed:
Fix typo in `re2c_manpage_php`

skvadrik reviewed (changes requested):
Thanks for the port!

A minor thing, but can you put all PHP entries sorted alphabetically, between OCaml and  Python?

Also, please rebase on top of master.

This is a partial review. I'll do the full review as soon as I can, but I broke my arm a few days ago, and I'm slower than usual.

Jeckerson committed:
Sort PHP code implementation alphabetically

Jeckerson committed:
Refactor 01_fill.re

Jeckerson commented:
@skvadrik I adjusted alphabetically the implementation. There are still work in examples/ and I'll need your help in some already made examples, especially with `null` (`[\x00]`) value. As in PHP it must be generated like that: `case "\0":` and not `case 0x00:`.

skvadrik commented:
> I adjusted alphabetically the implementation.

> There are still work in examples/ and I'll need your help in some already made examples, especially with null ([\x00]) value. As in PHP it must be generated like that: case "\0": and not case 0x00:.

I think you should use `re2c:yych:literals = hex;`, similar to Python that also has no character type (single-quoted literals are strings, as in PHP).

skvadrik commented on include/syntax/php:
```diff
@@ -0,0 +1,381 @@
+single_quoted_strings = 1;
+indentation_sensitive = 0;
+wrap_blocks_in_braces = 1;
+
```
I've added a new configuration `special_escapes` that needs to go here, see other syntax files.

skvadrik commented on examples/php/01_basic.re:
```diff
@@ -0,0 +1,16 @@
+<?php
+// re2php $INPUT -o $OUTPUT
```
I think it should always be the first line of the file (as expected by run_tests.py).

Jeckerson replied:
Impossible, as `<?php` must be on 1st line.

skvadrik replied:
I see. Then I suggest adding some sed magic in `examples/php/__run_all.sh` that will cut off the comments in the temporary file (this one and the one autogenerated by re2php).

skvadrik commented on include/syntax/php:
```diff
@@ -0,0 +1,381 @@
+supported_api_styles = ["free-form"];
+supported_code_models = ["loop-switch"];
+supported_targets = ["code", "dot"];
+supported_features = ["nested-ifs", "bitmaps", "tags", "captures", "captvars"];
```
For all supported and unsupported features there must be a test. Tags, captures and captvars are covered by examples, and for the rest see tests in `test/codegen/<language>`.

skvadrik reviewed (changes requested):
A few more comments.

skvadrik commented:
So I downloaded the patch and tried to build some of the problematic examples, but it seems that the .php files are not generated from the current syntax file (it was copied from javascript without modification). Please let me know when you update it (or if you have any questions).

Jeckerson committed:
Fix PHP syntax: loop-switch model, const_global, code:continue

- code:loop: emit label for the while loop (yyl: while (true) { ... })
- code:continue: use goto label instead of break/continue N
  avoids PHP warnings about 'continue 2 targeting switch'
- code:const_global: use define() instead of variable assignment
  so YYMAXFILL and similar constants are accessible inside functions
- supported_features: remove tags/captures (PHP $-prefix incompatible)

Jeckerson committed:
Regenerate bootstrap PHP files

Jeckerson committed:
Add PHP support to test runner and --lang option

Jeckerson committed:
Add PHP codegen tests

Jeckerson committed:
Fix and update PHP examples

- Fix line order: // re2php line must come first for test runner detection
- Remove JS copies: headers/, includes/, reuse/, state/, submatch/, fill/02_fill.*
- Rewrite fill/01_fill.re as proper PHP (uses goto lex_top for restart)
- Fix 04_fake_sentinel.re: YYPEEK must not include semicolon (template adds it)
  and use ord($str[$cur]) not ord($cur) to convert char to int
- Fix 02_bounds_checking.re: add yyfill:parameter = 0 to avoid YYFILL(1) call
- Regenerate all .php output files with updated re2php

Jeckerson committed:
fix: regenerate bootstrap without timestamp, restore swift support

Regenerate bootstrap/src/options/parse_opts.cc using --no-generation-date
to match upstream format (no timestamp, relative path from build dir).
Restore swift detection (re2swift) and substitution that were accidentally
dropped when adding PHP support. Add swift extension to clean_test_tree().
